### PR TITLE
Drop in Unidata numpy workshop notebooks

### LINKED ISSUE #69: Kernel spec in template.ipynb
After running a jupyterbook build on the current main version of Pythia Foundations, I got a warning that the kernel spec that is in template.ipynb was not found:

> jupyter_client.kernelspec.NoSuchKernel: No such kernel named conda-env-pythia-book-dev-py

Currently, the kernelspec is set as follows in template.ipynb:

> "kernelspec": {
   "display_name": "Python [conda env:pythia-book-dev]",
   "language": "python",
   "name": "conda-env-pythia-book-dev-py"

Our other notebooks currently use "python3" for the kernel, and doing so for template.ipynb makes for a warning-free jupyter book build. 

Yet, our instructions mention creating and activating the "pythia-book-dev" environment ...

### PR DESCRIPTION
**Pull request by dcamron**
These are dropped in from [this directory](https://github.com/Unidata/python-training/tree/master/pages/workshop/NumPy) of NumPy notebooks from Unidata's workshop materials with exercises modified and branding removed. Otherwise, so far content un-modified.

Notes: 
 * Multiple notebooks, not sure about scope/content crossover and if this a good example of a larger unit in development
 * Not unique to this PR, but need to pull out code comments and create more full narratives

- [X] Closes #69

### CODE REVIEW

**Review by clyne — COMMENTED** [review]
test review. Ignore

**Comment by mgrover1** [comment]
These notebooks look great! I like the use of "synthetic weather data" which helps illustrate geoscience related data/plots, while also showing the user how to utilize numpy. The progression between the three notebooks is also helpful.

File: core/numpy/numpy-basics.ipynb [hunk 1]
```diff
@@ -0,0 +1,783 @@
+{
```
Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>Perhaps include an example here <em>trying</em> to do arithmetic with different shaped/sized arrays, and show the error.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>My personal style preference, which we can certainly debate:</p><p>I find the extra whitespace in these expression distracting and harder to read.</p><p> </p><p>I would write this as</p><pre class="ql-syntax" spellcheck="false">(0, 2*np.pi + np.pi/4, np.pi/4)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>It might help to add an extra example here to show that you get exactly the same results by doing this:</p><pre class="ql-syntax" spellcheck="false">degrees = t/np.pi * 180.
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>Add a comment here about the <code>.reshape()</code>  method</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>This is a neat graphic and seems really helpful for explaining how indexing works! </p><p> </p><p>Is it possible to modify the graphic so the array contents are consistent with the example (rather than all 9s)?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>I like the size and scope of this first lesson. As a general rule, I think each notebook should conclude with some sort of statement about "now you're ready for the next lesson, ..." to give readers some idea of what's coming next.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='62'/>

**Comment by brian-rose** [comment]
I just enabled [ReviewNB](https://app.reviewnb.com) on this repository, and I'm trying it out for more fine-grained comments on @dcamron's notebooks.

If others want to give it a try, I think you need to first log in to ReviewNB using your GitHub credentials.

File: core/numpy/numpy-intermediate.ipynb [hunk 2]
```diff
@@ -0,0 +1,441 @@
+{
```
Comment (brian-rose) [inline core/numpy/numpy-intermediate.ipynb:1]: <p>Typo, replace "some" with "sum"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (brian-rose) [inline core/numpy/numpy-intermediate.ipynb:1]: <p>This is a cool example!</p><p> </p><p>I suggest writing down the equation that we're solving here, which will be familiar to anyone who's done some fluids (both also understandable for others):</p><p> </p><pre class="ql-syntax" spellcheck="false">$$ \text{advection} = -\vec{v} \cdot \nabla T $$
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (brian-rose) [inline core/numpy/numpy-intermediate.ipynb:1]: <p>Add a comment about the <code>.dstack()</code>  method</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (brian-rose) [inline core/numpy/numpy-intermediate.ipynb:1]: <p>Add a note / crossreference about our matplotlib tutorials</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

File: core/numpy/numpy-broadcasting.ipynb [hunk 3]
```diff
@@ -0,0 +1,840 @@
+{
```
Comment (brian-rose) [inline core/numpy/numpy-broadcasting.ipynb:1]: <p>General comment on this section:</p><p> </p><p>I think it's great, but it's hard to avoid the fact that broadcasting in NumPy is <strong><em>complicated</em></strong>. I suggest that we add caveats both at the beginning and end of this section. We should point readers toward labeled data operations with xarray as a simpler and more elegant solution for broadcasting that exist on different physical axes.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

Comment (brian-rose) [inline core/numpy/numpy-broadcasting.ipynb:1]: <p>replace "as a loop" with "without an explicit loop"?</p><p> </p><p>(unless I misunderstood something)</p><p> </p><p>Also fix typo: "condsider" --&gt; "consider"</p><p> </p><p>Also, it is possible to improve the formatting of this table display? When I view the rendered version, the first 2 columns are squashed together.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

Comment (brian-rose) [inline core/numpy/numpy-broadcasting.ipynb:1]: <p>"This means you cannot accidentally modify the original data"</p><p> </p><p>I think this is closer to what you mean here.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='51'/>

Comment (brian-rose) [inline core/numpy/numpy-broadcasting.ipynb:1]: <p>In the rendered book, the inline equation doesn't render properly.</p><p> </p><p>Maybe replace with equation block rather than inline.</p><p> </p><p>Also for consistency, add an equation for 1st derivative above.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='52'/>

Comment (brian-rose) [inline core/numpy/numpy-broadcasting.ipynb:1]: <p>This part is all new to me, I've never heard of <code>strides</code>.</p><p> </p><p>I wonder if this part is out of scope for a foundational set of numpy lessons.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='63'/>

Comment (brian-rose) [inline core/numpy/numpy-broadcasting.ipynb:1]: <p>Same comment here as I made above about broadcasting: I think it would help to point readers toward xarray for more intuitive and robust indexing of labeled multidimensional data.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='87'/>

File: core/numpy/numpy-basics.ipynb (hunk 1 above)
Comment (mgrover1) [reply core/numpy/numpy-basics.ipynb:1]: <p>This is a great point - an "Up next" portion</p>

Comment (mgrover1) [inline core/numpy/numpy-basics.ipynb:1]: <p>Maybe add difference between list vs. array?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (mgrover1) [inline core/numpy/numpy-basics.ipynb:1]: <p>Consider adding a comparison between the list from the cell above and how <code>arange</code>  and <code>linspace</code>  produce the same thing with shorter syntax</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (mgrover1) [inline core/numpy/numpy-basics.ipynb:1]: <p>Add in specific geoscience example of where this would be helpful - perhaps a graphic would help with this?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

File: core/numpy/numpy-broadcasting.ipynb (hunk 3 above)
Comment (mgrover1) [inline core/numpy/numpy-broadcasting.ipynb:1]: <p>Provide context as to what this would look like without broadcasting (loop)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (mgrover1) [inline core/numpy/numpy-broadcasting.ipynb:1]: <p>More of an explanation of what this is doing would be helpful</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (mgrover1) [inline core/numpy/numpy-broadcasting.ipynb:1]: <p>I like the use of this example!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='69'/>

Comment (mgrover1) [inline core/numpy/numpy-broadcasting.ipynb:1]: <p>Similar to previous notebook, adding in "what's up next" section would be helpful! Great content!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='88'/>

File: core/numpy/numpy-intermediate.ipynb (hunk 2 above)
Comment (mgrover1) [inline core/numpy/numpy-intermediate.ipynb:1]: <p>Add bullet points with additional materials - maybe some things from here? https://numpy.org/doc/stable/user/quickstart.html</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

File: core/numpy/numpy-broadcasting.ipynb (hunk 3 above)
Comment (dopplershift) [reply core/numpy/numpy-broadcasting.ipynb:1]: <p>No, you very much can accidentally modify your data.</p><pre class="ql-syntax" spellcheck="false">import numpy as np<span style="font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; background-color: white; color: rgba(0, 0, 0, 0.87); letter-spacing: 0.01em;">
</span>a = np.arange(9).reshape(3, 3)
b = a[1]
b[1] = 100
print(a)
</pre><p>prints</p><pre class="ql-syntax" spellcheck="false">[[ &nbsp;0 &nbsp; 1 &nbsp; 2]
&nbsp;[ &nbsp;3 100 &nbsp; 5]
&nbsp;[ &nbsp;6 &nbsp; 7 &nbsp; 8]]
</pre>

Comment (dopplershift) [reply core/numpy/numpy-broadcasting.ipynb:1]: <p>This particular use of <code>as_strided</code>  is definitely advanced stuff and could very well be left out. The general concept of strides though of part of even Python's own slicing:</p><pre class="ql-syntax" spellcheck="false">l = list(range(15))
l[::3]
</pre><p>gets the list with a stride of 3. Strides themselves I consider a core concept.</p>

File: core/numpy/numpy-basics.ipynb (hunk 1 above)
Comment (dopplershift) [reply core/numpy/numpy-basics.ipynb:1]: <p>I don't think we have the source around for this anywhere. One thing that's nice about the 9's is that it makes it easy to distinguish from the outside numbers. Might need some color or something to otherwise make that easier if we change the 9's.</p>

File: core/numpy/numpy-broadcasting.ipynb (hunk 3 above)
Comment (dopplershift) [reply core/numpy/numpy-broadcasting.ipynb:1]: <p>Broadcasting is definitely advanced numpy-fu, and linking to xarray is definitely good. </p>

**Comment by clyne** [comment]
In the spirt of streamlining reviews I am removing myself as a reviewer. It looks like there are plenty of comments from others already.

**Comment by clyne** [comment]
Hmm. I don't seem to be able to remove myself...

**Comment by brian-rose** [comment]
> Hmm. I don't seem to be able to remove myself...

Yeah, AFAIK github doesn't let you do that. Maybe @dopplershift knows differently? You can always just ignore the review request though.

**Comment by clyne** [comment]
I plan to ignore it, and proclaim my ignorance in public :-)

**Comment by clyne** [comment]
@dcamron just checking in. Do you think this is still doable for Wednesday? Sorry to nag :-(

**Comment by dcamron** [comment]
> @dcamron just checking in. Do you think this is still doable for Wednesday? Sorry to nag :-(

Yep!

File: core/numpy/numpy-basics.ipynb (hunk 1 above)
Comment (dcamron) [reply core/numpy/numpy-basics.ipynb:1]: <p>Can't do this with <code>black</code></p>

**Review by mgrover1 — APPROVED** [review]
this looks great - thank you for making all those changes!

**Comment by brian-rose** [comment]
Awesome, I will get to this tonight or tomorrow morning.

File: core/numpy/intermediate-numpy.ipynb [hunk 4]
```diff
@@ -0,0 +1,674 @@
+{
```
Comment (brian-rose) [inline core/numpy/intermediate-numpy.ipynb:1]: <p>This seems to refer to an exercise that we don't have here.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (brian-rose) [inline core/numpy/intermediate-numpy.ipynb:1]: <p>Add a note here like "Recall that <code>axis=-1</code>  refers to the last axis, which in this case means the sum of x and y."</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (brian-rose) [inline core/numpy/intermediate-numpy.ipynb:1]: <p>This section is pretty long. Is there a way to break it up with titled subsections?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (brian-rose) [inline core/numpy/intermediate-numpy.ipynb:1]: <p>This material is really good. We <em>could</em> introduce <code>np.where()</code>  here to show how to implement the same logic with a numpy-native method. The point would mostly be to demonstrate that numpy has a rich library of methods for manipulating arrays that is well worth exploring. Not necessary, but that was my thought on reading this.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='55'/>

File: core/numpy/numpy-basics.ipynb [hunk 5]
```diff
@@ -0,0 +1,975 @@
...
```
Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>This callout is rendering oddly. Can it be made consistent with the example "Warning" alert in the template?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>proves ---&gt; provides</p><p>(I think that's what you meant)</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>"... and try rounding the result to three decimal places"</p><p> </p><p>(just in case the operation isn't obvious to the reader)</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='48'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>Same comment as above -- looks funny in the rendered version.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='85'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>ever --&gt; every</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='86'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>I suggest breaking this up with a new subheading here, "Slicing along multiple dimensions" or something like that.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='93'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>I think this section can use a little more description of what the slice notation means in words, like "Here we are asking for every column (<code>:</code>) from rows zero and one (<code>0:2</code>)"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='95'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>"Here we are asking for every row (<code>:</code>)  from every second column beginning from column zero (<code>::2</code>)"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='97'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>"This means everything from row zero"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='100'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>"Everything along the last index of the final dimension"</p><p> </p><p>It may be helpful to state that *in this 3D example*, this is entirely equivalent to </p><pre class="ql-syntax" spellcheck="false">c[:,:,-1]
</pre><p>but is useful when you don't know (or care) about the number of dimensions.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='101'/>

Comment (brian-rose) [inline core/numpy/numpy-basics.ipynb:1]: <p>Fix capitalization: Pandas, Xarray</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='103'/>

File: core/numpy/numpy-broadcasting.ipynb [hunk 6]
```diff
@@ -0,0 +1,921 @@
...
```
Comment (brian-rose) [inline core/numpy/numpy-broadcasting.ipynb:1]: <p>It would be great to break this section up into multiple subsections.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (brian-rose) [inline core/numpy/numpy-broadcasting.ipynb:1]: <p>"iterating over all the points" (typo)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='64'/>

Comment (brian-rose) [inline core/numpy/numpy-broadcasting.ipynb:1]: <p>Fix capitalization: Pandas, Xarray</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='97'/>

**Review by brian-rose — COMMENTED** [review]
I left detailed comments on ReviewNB. One additional comment: it would be great to update the [landing page of the numpy section](https://foundations.projectpythia.org/core/numpy.html) to have a brief description of the topics covered in the three notebooks.

(That can be a separate PR or included here, either is fine)

File: core/numpy/intermediate-numpy.ipynb (hunk 4 above)
Comment (dcamron) [reply core/numpy/intermediate-numpy.ipynb:1]: <p>Good catch, I removed that.</p>

File: core/numpy/numpy-basics.ipynb (hunk 5 above)
Comment (dcamron) [reply core/numpy/numpy-basics.ipynb:1]: <p>These are often stylized this way in-line, check out the <a href="https://pandas.pydata.org/docs/getting_started/overview.html" target="_blank">pandas docs</a> and <a href="http://xarray.pydata.org/en/stable/" target="_blank">xarray docs</a>. We can decide on a consistent stylization for non sentence starters if-needed and I don't think anyone would be too mad if I did capitalize these. Either way.</p>

File: core/numpy/intermediate-numpy.ipynb (hunk 4 above)
Comment (dcamron) [reply core/numpy/intermediate-numpy.ipynb:1]: <p>I think that's valuable! It also extends to potentially common uses within xarray, but maybe I'll open an issue on covering it in another notebook (separating out intermediate content?) or extending this notebook sometime in the future</p>

File: core/numpy/numpy-basics.ipynb (hunk 5 above)
Comment (dcamron) [reply core/numpy/numpy-basics.ipynb:1]: <p>Yep just left in something I was messing around with on accident</p>

**Comment by dcamron** [comment]
I have to step out for a bit, but I think the notebook feedback has all been incorporated. I agree about expanding on the NumPy chapter page (and think that's valuable for all of our content so far) and will tackle that here or in a new PR later today or first thing in the morning.

File: core/numpy/numpy-basics.ipynb (hunk 5 above)
Comment (brian-rose) [reply core/numpy/numpy-basics.ipynb:1]: <p>No big deal! There are more important things to get mad about.</p>

**Review by brian-rose — APPROVED** [review]
Looks great! Thanks for making the revisions. The added subheadings are very welcome.

I suggest we merge this and open a separate PR to update the numpy landing page, in order to keep the reviews streamlined.

**Comment by brian-rose** [comment]
Note to self (or someone else): we still need to update the [Numpy landing page](https://foundations.projectpythia.org/core/numpy.html) once this PR is merged.

**Comment by mgrover1** [comment]
Looks good to me! Ready to merge!

**Comment by brian-rose** [comment]
There's a little conflict with some notebook metadata. I will fix and merge.

**Comment by mgrover1** [comment]
> There's a little conflict with some notebook metadata. I will fix and merge.

Ope good catch - thanks!
